### ISSUE TITLE
cacheHit is false when loading from cache on re-renders

### ISSUE THREAD

**Issue by WayneEllery** [nearform/graphql-hooks#428]
### Package

`graphql-hooks`

### Environment

- `graphql-hooks` version:
  
- `graphql-hooks-ssr` version: 4.3.0
- `graphql-hooks-memcache` version: 1.3.0
- `react` version: 16.12.0
- Browser: Chrome 79.0.3945.79

### Description

Currently on re-renders `cacheHit` is not updated and stays as `false` even when data is loaded from cache. 

### How to reproduce

[Codesandbox](https://codesandbox.io/s/github/WayneEllery/graphql-hoocks-cache-hit/tree/master/?fontsize=14)
[Github](https://github.com/WayneEllery/graphql-hoocks-cache-hit)

```JavaScript
import React, { useState } from 'react';
import { useQuery } from 'graphql-hooks';

const COUNTRIES_QUERY = `query Countries {
  countries {
    code
    name
  }
}`;

const ChildComponent = () => {
  const [count, setCount] = useState(0);
  const { loading, error, cacheHit } = useQuery(COUNTRIES_QUERY);

  if (loading || error) {
    return <></>;
  }

  console.log({ cacheHit });

  return (
    <>
      <p>You clicked {count} times</p>
      <button onClick={() => setCount(count + 1)}>Click me</button>
    </>
  );
};

export default ChildComponent;
```

When `Click me` button is clicked `cacheHit` is logged as false even though in the network tab the query network call is only made on initial mount. I would expect `cacheHit` to be `false` on initial mount but would be `true` on future renders.

### Suggested solution (optional)

I had a quick look at the code and it might be related to this: 
```JavaScript
// NOTE: state from useReducer is only initialState on the first render
// in subsequent renders the operation could have changed
// if so the state would be invalid, this effect ensures we reset it back
const stringifiedCacheKey = JSON.stringify(cacheKey)
React.useEffect(() => {
  if (!initialOpts.updateData) {
    // if using updateData we can assume that the consumer cares about the previous data
    dispatch({ type: actionTypes.RESET_STATE, initialState })
  }
}, [stringifiedCacheKey]) // eslint-disable-line react-hooks/exhaustive-deps
```

https://github.com/nearform/graphql-hooks/blob/master/packages/graphql-hooks/src/useClientRequest.js#L102

Unfortunately `stringifiedCacheKey` doesn't change so state is not updated. `initialState. cacheHit ` is updated correctly but it's not updated in state:

```JavaScript
const isDeferred = initialOpts.isMutation || initialOpts.isManual
const initialCacheHit =
  initialOpts.skipCache || !client.cache ? null : client.cache.get(cacheKey)
const initialState = {
  ...initialCacheHit,
  cacheHit: !!initialCacheHit,
  loading: isDeferred ? false : !initialCacheHit
}
```

**Comment by bmullan91** [comment]
@WayneEllery The issue resides here: https://github.com/nearform/graphql-hooks/blob/master/packages/graphql-hooks/src/useQuery.js#L30-L32. We never call the `queryReq` from `useClientRequest` again since the option did not change between renders. I'm not sure if this is an issue, what are you using `cacheHit` for?

**Comment by WayneEllery** [comment]
@bmullan91 Thanks for getting back to me. I've been using it for 2 reasons:

- We need to fire a tracking event with the data when data is loaded from the api call instead of from cache. Eg ```
const { loading, error, data, refetch, cacheHit } = useQuery(QUERY)
if (cacheHit) {
  track(data);
}
```
- The second reason is for debugging and wanted to output to the console. I needed to know when data is loaded from an api call and when it's loaded from the cache. It the readme it specifically says that this property is useful for this case but it isn't reliable so my only option I've got for this currently is to inspect the network tab.
